### PR TITLE
Find and replace previous PMG meeting files

### PR DESCRIPTION
**Pull request by waracci**
in dev 
```
docker-compose run --rm web python utils/find_replace.py --csv data/input.csv --out data/output.csv --logfile data/output.log
```

configuring prod (via ansible)
```
dokku docker-options:add pmg run "-v /mnt/...snapshotdir...:/data
```

in prod
```
dokku --rm run web python utils/find_replace.py --csv data/input.csv --out data/output.csv --logfile data/output.log
```

### CODE REVIEW

File: utils/find_replace.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        row_url = row["url"]
+        meeting_id = row["meeting_id"]
+        meeting = find_meeting_by_id(meeting_id)
+        if not meeting:
+            print("Could not find meeting {}".format(meeting_id))
+            continue
+        new_file_url = ""
+        try:
+            with app.app_context():
+                file = upload_file(row["filesystem_path"])
```
Comment (waracci) [inline utils/find_replace.py:51]: copy file path to tmp dir

File: utils/find_replace.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                new_file_url = file.url
+                event_file = EventFile(event=meeting, file=file)
+        except FileNotFoundError:
+            continue
```
Comment (waracci) [inline utils/find_replace.py:55]: log file path, committee_id error for exceptional circumstances.

File: utils/find_replace.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    write_to_csv(args.out, process_and_upload_files(args))
+
+
+if __name__ == "__main__":
+    print("Starting the find and replace file links")
+    parser = argparse.ArgumentParser(description=__name__)
+    parser.add_argument(
+        "--csv", required=True, help="path to csv file with extracted urls"
+    )
+    parser.add_argument(
```
Comment (waracci) [inline utils/find_replace.py:99]: Add a log file path

Comment (waracci) [reply utils/find_replace.py:99]: check volume name

Comment (waracci) [reply utils/find_replace.py:124]: done

File: utils/find_replace.py (hunk 2 above)
Comment (waracci) [reply utils/find_replace.py:77]: done

File: pmg/models/s3_upload.py [hunk 4]
```diff
@@ -80,7 +80,9 @@ def upload_file(self, path, key):
         except Exception as e:
             logger.error("Cannot upload file to S3. Removing file from disk.")
             # remove file from disc
-            os.remove(path)
+            # We kinda need to deactivate this during this process
+            #  because if something happens during upload, the file will be deleted
+            # os.remove(path)
```
Comment (jbothma) [inline pmg/models/s3_upload.py:85]: revert this

File: utils/find_replace.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    return meeting
+
+
+def upload_file(path):
+    with tempfile.TemporaryDirectory() as temp_dir:
+        temp_path = os.path.join(temp_dir, os.path.basename(path))
+        shutil.copyfile(path, temp_path)
+        file = File()
+        file.from_file_blob(temp_path)
+        file.file_path = path
```
Comment (jbothma) [inline utils/find_replace.py:50]: should we be overwriting this property? does it not get sit by from_file_blob?

File: pmg/models/s3_upload.py (hunk 4 above)
Comment (waracci) [reply pmg/models/s3_upload.py]: done

File: utils/find_replace.py (hunk 5 above)
Comment (waracci) [reply utils/find_replace.py]: No, we shouldn't. Removed.

File: utils/index.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import hashlib
+import os
+from collections import namedtuple
+
+
+def get_backup_file_path(path):
+    with open(path, 'r') as f:
+        reader = csv.reader(f)
+        for row in reader:
+            yield row[0], row[2]
```
Comment (jbothma) [inline utils/index.py:21]: https://docs.python.org/3/library/csv.html#csv.DictReader is your friend - next time - no need to change now

File: utils/index.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        # TODO: check if event info is already in csv
+        filename = os.path.basename(event_info[1])
+        # escape characters such as []
+        dir = f"**/{glob.escape(filename)}"
+        files_found = glob.glob(dir, recursive=True)
+        if not len(files_found):
+            continue
+        yield from compare_and_yield_row(
+            files_found,
+            hashes, csv_row,
```
Comment (jbothma) [inline utils/index.py:61]: why is this level aware of hashes? we don't care about duplicates _across_ rows, only multiple matches to the path on one row

File: utils/index.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            event_info
+        )
+
+
+def write_to_csv(csv_path, rows):
+    with open(csv_path, 'w') as f:
+        writer = csv.writer(f)
+        writer.writerow(['meeting_id', 'url', 'filesystem_path', 'possible_duplicate'])
```
Comment (jbothma) [inline utils/index.py:69]: https://docs.python.org/3/library/csv.html#csv.DictWriter is your friend - next time - no need to change now

File: utils/index.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        try:
+            md5_hash = md5(file)
+        except IsADirectoryError:
+            continue
+        unique_hash.add(md5_hash)
+        if len(unique_hash) > 1:
+            possible_duplicate = True
+        if not hashes.get(md5_hash):
+            hashes[md5_hash] = file
+            yield csv_row(event_info[0], event_info[1], hashes[md5_hash], possible_duplicate)
```
Comment (jbothma) [inline utils/index.py:45]:
```suggestion
            yield csv_row(event_info[0], event_info[1], ",".join(hashes.values()), possible_duplicate)
```

Comment (jbothma) [reply utils/index.py:45]: but `hashes` should be initialised inside this function - the point is duplicates within a row.

File: utils/find_replace.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        row_url = row["url"]
+        meeting_id = row["meeting_id"]
+        filesystem_path = row["filesystem_path"]
+        meeting = find_meeting_by_id(meeting_id)
+        if not meeting:
+            logger.error("Could not find meeting {}".format(meeting_id))
+            continue
+        new_file_url = ""
+        try:
+            with app.app_context():
```
Comment (jbothma) [inline utils/find_replace.py:69]: Try moving this to main()

Comment (waracci) [reply utils/find_replace.py]: Done, works just fine when the app context is initialized on main.

**Pull request merged**